src/net_processing.cpp

```diff
@@ -3261,6 +3289,52 @@ bool PeerLogicValidation::SendMessages(CNode* pto, std::atomic<bool>& interruptM
             }
         }
 
+        // Check that outbound peers have reasonable chains
+        // GetTime() is used by this anti-DoS logic so we can test this using mocktime
+        int64_t time_in_seconds = GetTime();
+        if (!(state.m_protect_from_disconnect || pto->fInbound || pto->m_manual_connection || pto->fFeeler || pto->fOneShot) && state.fSyncStarted) {
+            // This is an outbound peer subject to disconnection if their chain
+            // lags behind ours (note: if their chain has more work than ours,
+            // we should sync to it, unless it's invalid, in which case we
+            // should find that out and disconnect from them elsewhere).
+            if (state.pindexBestKnownBlock != nullptr && state.pindexBestKnownBlock->nChainWork >= chainActive.Tip()->nChainWork) {
```

Comment:
In commit "Permit disconnection of outbound peers on bad/slow chains":

I don't see why the first and second branches of this if/else are distinct. It seems like you could unite them and drop unneeded checks with:

```c++
if (!state.m_header_with_required_work || (state.pindexBestKnownBlock && state.pindexBestKnownBlock->nChainWork >= state.m_header_with_required_work->nChainWork) {
    // Reset timeout and min required work if peer caught up to previous requirement.
    state.m_headers_chain_timeout = time_in_seconds + CHAIN_SYNC_TIMEOUT;
    state.m_header_with_required_work = chainActive.Tip();
    state.m_sent_getheaders_to_check_chain_sync = false;
} else if (time_in_seconds > state.m_headers_chain_timeout) {
...
```

Reply:
Hm, I'll think about this.  I haven't quite convinced myself yet that there are no edge-case differences between your simplification and my original patch but you may be right.